### PR TITLE
Add RequestError DB table and features

### PR DESCRIPTION
**Pull request by sumincho22**
[CLOUDBLD-8700](https://issues.redhat.com/browse/CLOUDBLD-8700)

Signed-off-by: Sumin Cho <sucho@redhat.com>

## Overview

> Any web service, during communication with the client, should provide a comprehensive response when something goes wrong, regardless of the source of the problem. That can help the client to build more reliable logic around the communication and react appropriately to the situation.
> 
> From the maintenance perspective, the information can help to collect information about frequency of one or another type of error and faster determine the actual cause of some problem.

**Source**: Serhii Tsymbaliuk | **Documentation**: ["Cachito Errors and Exceptions"](https://docs.google.com/document/d/1gfAp_dDCL9DRPGinhYo4hrVGog3haYT_pbibJ86v4IY/edit#)

In Serhii's documentation (linked above), there is an extensive list of client and server errors that relates to Cachito. With the application of these specific errors in the code, the proposal also calls for the addition of the request error table. By doing so, Cachito can record descriptive error logs in the database, which we can use to monitor, diagnose, and fix errors in more effective ways.

This implementation features the following changes in the web API:
> The `error_origin`, `error_type` and `error_message` fields need to be added to the response object of ‘Request’ for `/requests`, `/requests/<id>` and `/request-metrics` endpoints.

> The `/requests` and `/request-metrics` should be supplied with filters by `error_origin` and `error_type`. The filters can take multiple values:
> - `GET /api/v1/requests?error_origin=server`
> - `GET /api/v1/requests?error_type=FileAccessError,FilePermissionError`

> The classification allows extending statistics provided by `/api/v1/request-metrics/summary`.
The following metrics has to be included in the summary:
> - `client_errors` - number of requests failed with a repository error
> - `server_errors` - number of requests failed with a deployment, system or third-party error

## Implementation Checklist
- [X] Define all error types
- [X] Add RequestError DB table
- [X] Add `error_origin` and `error_type` to Cachito Web API (PATCH `/api/v1/requests/<id>`)
- [X] Rewrite `failed_request_callback` to recognize and send the error information.
- [X] Extend request metrics
- [X] Extend web API
- [x] Write new error handlers in `cachito.web.errors`
- [x] Register those error handlers in `cachito.web.app`
- [x] Migrate database
- [x] Apply new error types (worker and web errors)
- [x] Update corresponding unit and integration tests (worker and web errors)
- [x] Update API documentation

# Maintainers will complete the following section

- [ ] Commit messages are descriptive enough
- [ ] Code coverage from testing does not decrease and new code is covered
- [ ] OpenAPI schema is updated (if applicable)
- [ ] DB schema change has corresponding DB migration (if applicable)
- [ ] README updated (if worker configuration changed, or if applicable)

### CODE REVIEW

File: cachito/errors.py [hunk 1]
```diff
@@ -2,11 +2,11 @@
...
 
 
-class RequestErrorOrigin(Enum):
+class RequestErrorOrigin(str, Enum):
     """An Enum that represents the request error origin."""
 
-    client = 1
-    server = 2
+    client = "client"
+    server = "server"
```
Comment (lkolacek) [inline cachito/errors.py:9]: Can we make this change in first commit `Define all error types` ?

Comment (sumincho22) [reply cachito/errors.py:9]: Yep, I reorganized edits so that they belong to their respective commits.

Comment (lkolacek) [reply cachito/errors.py:9]: Thank you sir :)

**Comment by lkolacek** [comment]
@sumincho22 PR looks great to me so far! Is it ready for merging? Was it tested already? Why some error types were not used e.g. FilePermissionError and DatabaseError?

**Comment by sumincho22** [comment]
> @sumincho22 PR looks great to me so far! Is it ready for merging? Was it tested already? Why some error types were not used e.g. FilePermissionError and DatabaseError?

This PR is ready for merge. I also tested and passed all unit and integration tests. Some error types were not used because there were no errors corresponding to them. These unused errors could potentially be used for future cases.

**Comment by taylormadore** [comment]
Great job @sumincho22! I agree with @lkolacek that it looks really good. I've also pulled it locally to test the endpoints and found no issues so far.

**Comment by sumincho22** [comment]
@lkolacek @taylormadore Thank you guys for your feedback. Let me know if the new unit and integration tests I added seem valid. If you guys think it's all good, please leave me a review approval :)

**Comment by lkolacek** [comment]
@sumincho22  coverage slightly dropped but if our colleagues are OK with that so do I.

**Comment by sumincho22** [comment]
FYI the DB migration was validated by Adam.

**Review by lkolacek — APPROVED** [review]
Looks great. Thank you Sumin! :+1:

**Comment by sumincho22** [comment]
@taylormadore Cleared up the merging conflicts. This should be ready to merge now

**Review by taylormadore — APPROVED** [review]

**Review by ejegrova — APPROVED** [review]
great work!

**Comment by brunoapimentel** [comment]
Do we intend to completely remove the CachitoError class in a future PR? I guess this way, we could avoid that somebody kept using it as a mistake.

Awesome work here @sumincho22! :rocket: :rocket: :rocket:

**Comment by sumincho22** [comment]
> Do we intend to completely remove the CachitoError class in a future PR? I guess this way, we could avoid that somebody kept using it as a mistake.
> 
> Awesome work here @sumincho22! rocket rocket rocket

I agree. I think the whole point here was to avoid using CachitoError and be more clear about our errors. However, I was unsure if other files outside of the web and workers directories might still be using CachitoError, so I left it there just in case.

**Pull request merged**